### PR TITLE
ggml : do not sched_yield when calling BLAS

### LINKED ISSUE #291: ggml : add option for controlling work distribution across threads
See https://github.com/ggerganov/llama.cpp/pull/1507

And comment: https://github.com/ggerganov/llama.cpp/pull/1507#issuecomment-1605302021

> I guess we can extend ggml to be able to choose work chunk distribution method - either at compile time, or via a context parameter. We can factor out the range selections from the ggml forward implementations to make implementation more concise and extensible in the future

---

Another thing to be investigated is the usage of `sched_yield()` and potentially making it user configurable:

https://github.com/ggerganov/whisper.cpp/pull/1275/commits/09a6325de56856490ae9046bf0030ceedc04028a

### PR DESCRIPTION
**Pull request by ggerganov**
I think this should fix https://github.com/ggerganov/ggml/issues/291

### CODE REVIEW

**Comment by slaren** [comment]
I was very puzzled by the commend that the yield only works with with non-mul mat ops, and I think that the reason is that the compute for BLAS happens here, since `n_tasks = 1` when using BLAS:
https://github.com/ggerganov/llama.cpp/blob/23d9e5b6de7df234ffee9d0299c99530c0481a5d/ggml.c#L16536-L16548

So maybe this could be fixed by doing the yield if the next op is a mul mat with BLAS, or maybe just setting the `n_tasks` of the mul_mats to `n_threads`.

**Comment by ggerganov** [comment]
Good catch - I changed the matrix multiplication ops to use `n_tasks == n_threads` and yield on all `GGML_OP_MUL_MAT` ops regardless if BLAS is used.

**Review by slaren — APPROVED** [review]
I don't see any issues with the changes. For me, the performance both with and without OpenBLAS is the same as in master.

File: ggml.c [hunk 1]
```diff
@@ -16556,14 +16539,18 @@ static thread_ret_t ggml_graph_compute_thread(void * data) {
         } else {
             // wait for other threads to finish
             const int last = node_n;
+
+            const bool do_yield = last < 0 || cgraph->nodes[last]->op == GGML_OP_MUL_MAT;
```
Comment (gwjr) [inline ggml.c:16543]: Does this assume that GGML_OP_MUL_MAT is the only op that calls into blas?  If so, that isn't right: see ggml_compute_forward_out_prod_f32().

Comment (ggerganov) [reply ggml.c:16543]: It probably should be added, but I have a benchmark just for `GGML_OP_MUL_MAT` at the moment, so cannot verify it atm. If you observe that it improves the performance, we can add it

**Pull request merged**